Add SSH key support (cont. for #5154)

bcho opened the pull request:
###### Before submitting the PR, please make sure you do the following

- [ ] If you're attempting to fix a translation issue, please submit your changes to our [POEditor project](https://poeditor.com/join/project/ynjQmwSsGh) instead of opening a PR.

### Tests
-  [ ] Run the frontend and integration tests with  `yarn lint && yarn flow && yarn test`)
-  [x] If there are changes to the backend codebase, run the backend tests with `lein test && lein lint && ./bin/reflection-linter`

-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

------------------------

This pull request is the test follow up for #5154:

- added sshd as mock sssh server
- added ssh connection related test cases (password, key-based challenge)

fstrube committed:
Adding SSH key support

* added ssh-add-key utility function to keep code clean
* enabled the commented-out tunnel-private-key-file-name field on the connection form
* SSH password is no longer required, because you can connect with either a password or a key
* fixed bug with using :tunnel-host as the JDBC host (it should always be localhost when we're tunneling)

TODO

* write tests
* add support for passphrases on keys

bcho committed:
Add SSH connect testcases

- added `org.apache.sshd/sshd-core` as mock sshd server
- refine ssh utils code style
- add ssh connection test cases

bcho commented:
CLA signed! Please have a look @camsaul

bcho commented on project.clj:
```diff
@@ -72,6 +72,7 @@
                  it.unimi.dsi/fastutil]]
    [com.draines/postal "2.0.3"]                                       ; SMTP library
    [com.jcraft/jsch "0.1.55"]                                         ; SSH client for tunnels
+   [org.apache.sshd/sshd-core "2.3.0"]                                ; SSH test server
```
This is a test dependency, I didn't find a proper way to put it cc @camsaul

camsaul replied:
The `:dev` profile is probably the best place to put it

bcho replied:
Updated, PTAL @camsaul

bcho commented on test_resources/ssh_test:
```diff
@@ -0,0 +1,39 @@
+-----BEGIN RSA PRIVATE KEY-----
```
these keys were generated by `ssh-keygen -t rsa  -m PEM`

flyingmachine reviewed (changes requested):
This is great, thank you! The comments are mostly minor requests for changes :) There is one larger question for @mazameli about how to handle the SSH settings in the UI.

bcho committed:
Move ssh test fixtures to `test_resources/ssh`

bcho commented:
That's wired, I don't know why tests keep failing (no `more-of`). But last time I tested, it worked on circleci... @flyingmachine any idea? (The last time I wrote clojure was 2 years ago, I don't know is anything changed in that time, thanks in advance!)

bcho commented:
> There is one larger question for @mazameli about how to handle the SSH settings in the UI.

I don't have any suggestions here since I don't use metabase very often. But I think [sequel pro](https://www.sequelpro.com/) provides some great examples on how to handle SSH login (with or without ssh key pass phrase).

flyingmachine committed:
update tests to use clojure.test directly

we're moving away from the expectations library, and `expectations` namespace now is really a kind of facade over clojure.test. This is why the `more-of` macro wasn't available.

flyingmachine committed:
fully convert to deftest

flyingmachine committed:
clean ns

flyingmachine commented on src/metabase/util/ssh.clj:
```diff
@@ -52,13 +61,12 @@
     :type         :password
-    :placeholder  "******"
-    :required     true}
+    :placeholder  "******"}
```
@mazameli would love your thoughts here :)

flyingmachine commented:
@bcho I've updated the tests so that they pass now :) The issue was with how our testing approach has evolved to move from `expectations` to `clojure.test`. Thank you for your contribution!

Now we're just waiting on feedback on how to handle this in the UI.

flyingmachine committed:
remove logging statement

mazameli commented:
@flyingmachine If I'm understanding correctly the the password vs. the SSH key are mutually exclusive, then I think the cleanest solution would be to ask which auth method the user wants to use, and then making the password or SSH key field visible and required based on the selection. (I've indented these SSH settings in the mocks to clarify that they're sub-settings of the SSH-tunnel toggle, but that's a nicety, not a requirement.)

**Default:**

![image](https://user-images.githubusercontent.com/2223916/66356982-f11e5400-e921-11e9-8819-7e836aa4155c.png)

**If SSH key is selected:**

![image](https://user-images.githubusercontent.com/2223916/66356993-fd0a1600-e921-11e9-88e1-5c7a832045c8.png)

mazameli reviewed (changes requested):
I added some suggestions last week (above) for how to make the choice between using a key vs. a password clearer. @flyingmachine, can you let me know if the suggestions make sense to you? If so, then we'll need to ask @bcho to implement that change before we merge this.

mazameli commented:
@bcho, I confirmed with @flyingmachine that my [change requests](https://github.com/metabase/metabase/pull/10758#issuecomment-539252180) make sense to him. If you're able to make that change, we'd happily accept this PR. Thanks!

bcho commented:
OK I will have a look this weekend

On Thu, Oct 17, 2019 at 01:29 Maz Ameli <notifications@github.com> wrote:

> @bcho <https://github.com/bcho>, I confirmed with @flyingmachine
> <https://github.com/flyingmachine> that my change requests
> <https://github.com/metabase/metabase/pull/10758#issuecomment-539252180>
> make sense to him. If you're able to make that change, we'd happily accept
> this PR. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/metabase/metabase/pull/10758?email_source=notifications&email_token=AAPCGTXGUWUG4X4D2PYAW7LQO46PRA5CNFSM4ISEDGC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBND3DI#issuecomment-542784909>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPCGTWAGNXY3XI5EGH45VDQO46PRANCNFSM4ISEDGCQ>
> .
>

bcho commented:
@mazameli @flyingmachine 

I am not quite following what exactly the changes I have to do for the UI. Do I need to add a new input field (which render the radio widget like the mockup) in [here](https://github.com/metabase/metabase/blob/18da2fe56cfb6a2bc510ee6755f558a72c964e52/frontend/src/metabase/components/DatabaseDetailsForm.jsx#L153-L180)?

If it's true, I think it's better to split the tunnel rendering part out and render them like what the mockup do (because I don't find any examples for rendering fields like the suggested UI). Is that OK to do so?

flyingmachine commented:
@bcho thanks for having a look! As I understand it, the work would be to add This field:

```
How do you want to authenticate?
[ x ] Password
[    ] SSH key
```

Password and SSH key would be radio buttons. If you select password, then the UI would show a text input for the password: 

```
How do you want to authenticate?
[ x ] Password
[    ] SSH key

Password:
[ password input ]
```

If you select SSH key, then the UI would show a text field for the path to the key:

```
How do you want to authenticate?
[ x ] Password
[    ] SSH key

Path on the Metabase server to a SSH private key file:
[ text input for path ]
```

Does this help?

robdaemon commented:
I'm closing this in favor of #12548 - thank you for the patch! It got this feature going.
